pkg/registry/core/service/ipallocator/cidrallocator.go

```diff
@@ -0,0 +1,413 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipallocator
+
+import (
+	"fmt"
+	"net"
+	"net/netip"
+	"sync"
+	"time"
+
+	networkingv1alpha1 "k8s.io/api/networking/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	networkingv1alpha1informers "k8s.io/client-go/informers/networking/v1alpha1"
+	networkingv1alpha1client "k8s.io/client-go/kubernetes/typed/networking/v1alpha1"
+	networkingv1alpha1listers "k8s.io/client-go/listers/networking/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/util/iptree"
+	netutils "k8s.io/utils/net"
+)
+
+// CIDRAllocator maintains a Tree with the ServiceCIDRs containing an IPAllocator
+// on the Nodes. Since each allocator doesn't stored the IPAddresses because it reads
+// them from the informer cache, it is cheap to create and delete IPAllocators.
+// CIDRAllocator forwards the request to any of the internal allocators that has free
+// addresses.
+
+// Allocator implements current ipallocator interface using
+// ServiceCIDR and IPAddress API objects
+type CIDRAllocator struct {
+	client            networkingv1alpha1client.NetworkingV1alpha1Interface
+	serviceCIDRLister networkingv1alpha1listers.ServiceCIDRLister
+	serviceCIDRSynced cache.InformerSynced
+	ipAddressLister   networkingv1alpha1listers.IPAddressLister
+	ipAddressSynced   cache.InformerSynced
+	ipAddressInformer networkingv1alpha1informers.IPAddressInformer
+	queue             workqueue.RateLimitingInterface
+
+	internalStopCh chan struct{}
+
+	muTree sync.Mutex
+	tree   *iptree.Tree[*Allocator]
+
+	ipFamily api.IPFamily
+}
+
+var _ Interface = &CIDRAllocator{}
+
+// NewIPAllocator returns an IP allocator associated to a network range
+// that use the IPAddress objects to track the assigned IP addresses,
+// using an informer cache as storage.
+func NewCIDRAllocator(
+	client networkingv1alpha1client.NetworkingV1alpha1Interface,
+	serviceCIDRInformer networkingv1alpha1informers.ServiceCIDRInformer,
+	ipAddressInformer networkingv1alpha1informers.IPAddressInformer,
+	isIPv6 bool,
+) (*CIDRAllocator, error) {
+
+	family := api.IPv4Protocol
+	if isIPv6 {
+		family = api.IPv6Protocol
+	}
```

Comment:
this has to be revisited later as it depends on the allocator inicializaation , right now the allocators set the default family based on the service-cidr flag, but we are going to make the allocators configurable, so it may happen an apiserver can switch the default ip family ... I´ll add a TODO to not forget